src/topology/vector_bundle.lean

```diff
@@ -0,0 +1,151 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import topology.topological_fiber_bundle
+import analysis.calculus.deriv
+import linear_algebra.dual
+
+/-!
+# Topological vector bundles
+
+In this file we define topological vector bundles.
+
+The most important idea here is that vector bundles are named through their fibers.
+Let `B` be the base space. The collection of the fibers is a function `E : B → Type*` for which
+there is an appropriate instance on each `E x` and an instance of topological space over `Σ x, E x`.
+Naming conventions are essential to work with vector bundles this way.
+
+## Definitions
+
+* `nc_topological_vector_bundle R E F`  : topological vector bundle with non constant fiber. Here
+                                          `F` is a function `I → Type*` and `I` is an index type.
+* `topological_vector_bundle R E F`     : topological vector bundle constant fiber `F : Type*`.
+
+-/
+
+variables {B : Type*}
+
+/--
+`total_space E` is the total space of the bundle `Σ x, E x`. This type synonym is used to avoid
+conflicts with general sigma types.
+-/
+def bundle.total_space (E : B → Type*) := Σ x, E x
+
+instance {E : B → Type*} [inhabited B] [inhabited (E (default B))] :
+  inhabited (bundle.total_space E) := ⟨⟨default B, default (E (default B))⟩⟩
+
+/-- `bundle.proj E` is the canonical projection `total_space E → B` on the base space. -/
+def bundle.proj (E : B → Type*) : bundle.total_space E → B := λ y : (bundle.total_space E), y.1
+
+open bundle
+
+variables (R : Type*) (E : B → Type*) (F : Type*)
+[comm_semiring R] [∀ x, add_comm_monoid (E x)] [∀ x, semimodule R (E x)]
+
+/-- `bundle.dual R E` is the dual bundle. -/
+@[reducible] def bundle.dual := (λ x, module.dual R (E x))
+
+localized "notation E `ᵛ` R := bundle.dual R E" in bundle.dual
+
+section
+
+variables [topological_space B] [topological_space F] [topological_space (total_space E)]
+[add_comm_monoid F] [semimodule R F]
+
+@[nolint unused_arguments]
+instance {x : B} : has_coe_t (E x) (total_space E) := ⟨λ y, (⟨x, y⟩ : total_space E)⟩
+
+/-- Local trivialization for vector bunlde. -/
+@[nolint has_inhabited_instance]
+structure vector_bundle_trivialization extends bundle_trivialization F (proj E) :=
+(linear : ∀ x ∈ base_set, is_linear_map R (λ y : (E x), (to_fun y).2))
+
+instance : has_coe (vector_bundle_trivialization R E F) (bundle_trivialization F (proj E)) :=
+⟨vector_bundle_trivialization.to_bundle_trivialization⟩
+
+instance : has_coe_to_fun (vector_bundle_trivialization R E F) :=
+⟨_, λ e, e.to_bundle_trivialization⟩
+
+variables {I : Type*} (F' : I → Type*) [∀ i, topological_space (F' i)]
+[∀ i, add_comm_monoid (F' i)] [∀ i, semimodule R (F' i)]
+
+/-- Topological vector bundle with varying fiber. `nc` stands for non constant. -/
```

Comment:
Can you expand the docstring? From the current version, it is not clear what `nc_topological_vector_bundle` is (what are its arguments? is it the total space of the bundle, or an instance on this total space?)

Also, contrary to the case of manifolds (where the topology does not cary all the information, so it is necessary to have a definition which contains data, i.e., the charts), for a topological vector bundle once you have fixed the topology of everything then there is no further choice, so I think there should be no data in the definition, it should just be a Prop (requiring at every point the existence of a vector bundle trivialization).

Reply:
> Also, contrary to the case of manifolds (where the topology does not cary all the information, so it is necessary to have a definition which contains data, i.e., the charts), for a topological vector bundle once you have fixed the topology of everything then there is no further choice, so I think there should be no data in the definition, it should just be a Prop (requiring at every point the existence of a vector bundle trivialization).

I did it this way because in proofs you often want to say "take a vector bundle trivialization at point x" and this way it is immediate to do so, whereas the other way you'd need `classical.some` etc which is just clumsy. I mean I can see why philosophically it is more elegant as you propose, but I don't see what are the advantages practically. Is this a standard rule in mathlib though? If yes I'll just change it to a `class : prop` for coherence.

Reply:
The general rule is: if a property can be a Prop, then make it a Prop, because it will avoid all defeqness issues later on. I agree with you that using `classical.some` all over the place would be clumsy though, so what you can do is register definitions `fiber_index_at` and `trivialization_at` with `classical.some` (under the `nc_topological_vector_bundle` typeclass assumptions), prove their basic property `mem_trivialization_source`, and then no need to change anything to the rest of the formalization.